apps/hash-graph/lib/graph/src/store/postgres/traversal_context.rs

```diff
@@ -0,0 +1,188 @@
+            )
+            .await?
+            {
+                subgraph.insert_vertex(vertex_id, property_type);
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn read_entity_types_by_ids(
+        &self,
+        vertex_ids: impl IntoIterator<Item = &EntityTypeVertexId, IntoIter: Send> + Send,
+        subgraph: &mut Subgraph,
+    ) -> Result<(), QueryError> {
+        for vertex_id in vertex_ids {
+            for enttity_type in <Self as Read<EntityTypeWithMetadata>>::read_vec(
+                self,
+                &EntityTypeWithMetadata::create_filter_for_vertex_id(vertex_id),
+                Some(&subgraph.temporal_axes.resolved),
+            )
+            .await?
+            {
+                subgraph.insert_vertex(vertex_id, enttity_type);
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn read_entities_by_ids(
+        &self,
+        vertex_ids: impl IntoIterator<Item = &EntityEditionId, IntoIter: Send> + Send,
+        subgraph: &mut Subgraph,
+    ) -> Result<(), QueryError> {
+        for edition_id in vertex_ids {
+            for entity in <Self as Read<Entity>>::read_vec(
+                self,
+                &Filter::<Entity>::Equal(
+                    Some(FilterExpression::Path(EntityQueryPath::EditionId)),
+                    Some(FilterExpression::Parameter(Parameter::Uuid(
+                        edition_id.as_uuid(),
+                    ))),
+                ),
+                Some(&subgraph.temporal_axes.resolved),
+            )
+            .await?
+            {
+                subgraph.insert_vertex(
+                    &entity.vertex_id(subgraph.temporal_axes.resolved.variable_time_axis()),
+                    entity,
+                );
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct OntologyTraversalContext<R: Record> {
+    vertex_ids: HashSet<R::VertexId>,
+}
+
+impl<R: Record> Default for OntologyTraversalContext<R> {
+    fn default() -> Self {
+        Self {
+            vertex_ids: HashSet::new(),
+        }
+    }
+}
+
+impl<R: Record> OntologyTraversalContext<R>
+where
+    R::VertexId: Clone + Eq + std::hash::Hash,
+{
+    pub fn add_id(&mut self, vertex_id: R::VertexId) {
+        self.vertex_ids.insert(vertex_id);
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct EntityTraversalContext {
+    edition_ids: HashSet<EntityEditionId>,
+}
+
+impl EntityTraversalContext {
+    pub fn add_id(&mut self, edition_id: EntityEditionId) {
+        self.edition_ids.insert(edition_id);
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct TraversalContext {
+    data_types: OntologyTraversalContext<DataTypeWithMetadata>,
+    property_types: OntologyTraversalContext<PropertyTypeWithMetadata>,
+    entity_types: OntologyTraversalContext<EntityTypeWithMetadata>,
+    entities: EntityTraversalContext,
+}
+
+impl TraversalContext {
+    pub async fn load_vertices<C: AsClient>(
```

Comment:
Can we think of another name here? I'm not entirely sure I love `load_` but I don't have a good suggestion right now

Reply:
What do you think about `read_traversed_vertices`?

Reply:
[`867e3c4` (#2612)](https://github.com/hashintel/hash/pull/2612/commits/867e3c44e9aa0b7a494a5b9a296b983bf04fa88f)

Reply:
Are the vertices traversed yet? It sounds like they're candidates to be traversed, or in the queue, but haven't been until this function has completed. I'd also probably opt to just use normal nomenclature like `visit_x`

Reply:
Yes, the traversal context is populated when traversing the graph. `visit_x` sounds like you're going to traverse until `x`.

Reply:
I don't think that's true.. Let's say I have a super basic graph which is just a linked list `a -> b -> c`

If I'm doing a BFS/DFS (somewhat unimportant right now so ignore the post-order/pre-order whatever)

Then 
- I _visit_ `a`, and I look at its neighbors (`b`). I add `b` to the queue (note I have not _traversed_ `b`), I do whatever, and then
- I _visit_ `b`, and I look at its neighbors (`c`)...
- I _visit_ `c`...

In traversal you visit something when you read it, not when you add it to the queue. I don't think I've ever heard `visit_x` as being traversing _until_ `x`, is that how it's used in libraries like `serde`? 🤔 

(See ChatGPT's description for more support)
<img width="668" alt="image" src="https://github.com/hashintel/hash/assets/25749103/6f7813ca-af21-45b0-a833-cd035919b939">

Reply:
Ah I'm realising now that this operation doesn't actually add anything to the queue, it merely reads the records (which we don't need to do _during_ traversal?)

Reply:
Assuming we're not going to extend this to add more things to the explore 'queue', I don't feel strongly about the naming of this, happy to approve

Reply:
Yes, we don't need to read the records during the traversal. The purpose of this PR is to defer the reading of the vertices until after traversal so it's only required to look at the edges during traversal.

Later this week I'm going to change the `HashSet` to track which vertices were already visited so visiting the same vertex can be avoided (given that the resolve depths and time intervals are matching). The context, however, will only be used as the data which has been traversed, it will never hold any records of edges or vertices, which are not yet traversed so to directly answer your assumption: this will never be an "explore queue" but always an "explored set" 🙂